File: src/utils/api/tools/download.ts
/**
 * 工具模块 - 外部下载 + 文件名获取
 *
 * 对应后端 `tools_manager` 的 download_file / get_download_dir / list_downloads /
 * delete_download / fetch_filename action。
 */

import { TOOLS_ACTIONS, toolsManager } from './core'

export interface ExternalDownloadResult {
  path: string
  size: number
  file_name: string
}

export interface ExternalDownloadEntry {
  name: string
  size: number
  modified: number
}

export interface FetchFilenameResult {
  filename: string
  file_size: number
}

/** 外部下载高级设置（可选，全部为空时后端走全局下载配置） */
export interface DownloadFileSettings {
  /** 自定义请求 User-Agent（空 = 使用默认） */
  userAgent?: string
  /** 并发线程数（0/undefined = 使用全局配置） */
  maxThreads?: number
  /** 单文件分片数（0/1/undefined = 使用全局配置或单流） */
  chunkCount?: number
  /** 限速 bytes/秒（0/undefined = 不限速） */
  maxSpeed?: number
}

/** 新建外部下载任务 */
export function downloadFile(
  url: string,
  fileName: string,
  settings?: DownloadFileSettings,
): Promise<ExternalDownloadResult> {
  return toolsManager<ExternalDownloadResult>(TOOLS_ACTIONS.DOWNLOAD_FILE, {
    url,
    file_name: fileName,
    user_agent: settings?.userAgent?.trim() || undefined,
    max_threads: settings?.maxThreads || undefined,
    chunk_count: settings?.chunkCount || undefined,
    max_speed: settings?.maxSpeed || undefined,
  })
}

/** 获取下载目录（自定义或默认 .Molaunch/Download/） */
export function getDownloadDir(): Promise<string> {
  return toolsManager<string>(TOOLS_ACTIONS.GET_DOWNLOAD_DIR)
}

/** 列举已下载文件 */
export function listDownloads(): Promise<ExternalDownloadEntry[]> {
  return toolsManager<ExternalDownloadEntry[]>(TOOLS_ACTIONS.LIST_DOWNLOADS)
}

/** 删除已下载文件 */
export function deleteDownload(fileName: string): Promise<void> {
  return toolsManager<void>(TOOLS_ACTIONS.DELETE_DOWNLOAD, { file_name: fileName })
}

/** 从 URL 响应头获取文件名（Content-Disposition / URL 路径推断） */
export function fetchFilename(url: string): Promise<FetchFilenameResult> {
  return toolsManager<FetchFilenameResult>(TOOLS_ACTIONS.FETCH_FILENAME, { url })
}

File: src-tauri/src/commands/tools/types/download.rs
use serde::{Deserialize, Serialize};

/// 外部下载请求参数
#[derive(Debug, Serialize, Deserialize)]
pub struct DownloadFileParams {
    pub url: String,
    pub file_name: String,
    /// 自定义请求 User-Agent（为空时使用默认 UA）
    #[serde(default)]
    pub user_agent: Option<String>,
    /// 文件级并发线程数（覆盖全局下载配置）
    #[serde(default)]
    pub max_threads: Option<u32>,
    /// 单文件分片数（覆盖全局下载配置，0/1 = 单流）
    #[serde(default)]
    pub chunk_count: Option<u32>,
    /// 全局限速 bytes/秒（0 = 不限速，覆盖全局下载配置）
    #[serde(default)]
    pub max_speed: Option<u64>,
}

/// 删除已下载文件请求参数
#[derive(Debug, Serialize, Deserialize)]
pub struct DeleteDownloadParams {
    pub file_name: String,
}

/// 外部下载结果
#[derive(Debug, Serialize, Deserialize)]
pub struct ExternalDownloadResult {
    pub path: String,
    pub size: u64,
    pub file_name: String,
}

/// 已下载文件条目
#[derive(Debug, Serialize, Deserialize)]
pub struct ExternalDownloadEntry {
    pub name: String,
    pub size: u64,
    /// Unix 时间戳（秒）
    pub modified: u64,
}
